packages/consumption/src/modules/requests/incoming/IncomingRequestsController.ts

```diff
@@ -245,6 +272,150 @@ export class IncomingRequestsController extends ConsumptionBaseController {
+            }
+        }
+
+        if (fragmentsOfMustBeAcceptedItemsOfGroup.length !== 0) return fragmentsOfMustBeAcceptedItemsOfGroup;
+
+        return;
+    }
+
+    private extractRelationshipAttributeFragmentsFromAcceptedItemsOfGroup(requestItemGroup: RequestItemGroup, decideGroupParams: DecideRequestItemGroupParametersJSON) {
+        const fragmentsOfAcceptedItemsOfGroup: RelationshipAttributeFragment[] = [];
+
+        for (let i = 0; i < decideGroupParams.items.length; i++) {
+            const decideItemParams = decideGroupParams.items[i];
+            const item = requestItemGroup.items[i];
+
+            if (item instanceof RequestItem) {
+                const fragment = this.extractRelationshipAttributeFragmentFromAcceptedRequestItem(item, decideItemParams);
+                if (fragment) fragmentsOfAcceptedItemsOfGroup.push(fragment);
+            }
+        }
+
+        if (fragmentsOfAcceptedItemsOfGroup.length !== 0) return fragmentsOfAcceptedItemsOfGroup;
+
+        return;
+    }
+
+    private extractRelationshipAttributeFragmentFromMustBeAcceptedRequestItem(requestItem: RequestItem) {
+        if (requestItem.mustBeAccepted) {
+            return this.extractRelationshipAttributeFragmentFromRequestItem(requestItem);
+        }
+
+        return;
+    }
+
+    private extractRelationshipAttributeFragmentFromAcceptedRequestItem(requestItem: RequestItem, decideItemParams: DecideRequestItemParametersJSON) {
+        if (decideItemParams.accept) {
+            return this.extractRelationshipAttributeFragmentFromRequestItem(requestItem);
+        }
+
+        return;
+    }
+
+    private extractRelationshipAttributeFragmentFromRequestItem(requestItem: RequestItem) {
+        if (requestItem instanceof CreateAttributeRequestItem && requestItem.attribute instanceof RelationshipAttribute) {
+            const ownerIsEmptyString = requestItem.attribute.owner.toString() === "";
+            return {
+                owner: ownerIsEmptyString ? this.identity.address.toString() : requestItem.attribute.owner.toString(),
+                key: requestItem.attribute.key,
+                value: { "@type": requestItem.attribute.value.toJSON()["@type"] }
+            };
+        } else if (
+            (requestItem instanceof ReadAttributeRequestItem || requestItem instanceof ProposeAttributeRequestItem) &&
+            requestItem.query instanceof RelationshipAttributeQuery
+        ) {
+            const ownerIsEmptyString = requestItem.query.owner.toString() === "";
+            return {
+                owner: ownerIsEmptyString ? this.identity.address.toString() : requestItem.query.owner.toString(),
+                key: requestItem.query.key,
+                value: { "@type": requestItem.query.attributeCreationHints.valueType }
+            };
+        }
+
+        return;
+    }
+
+    private static containsDuplicateRelationshipAttributeFragments(fragments: RelationshipAttributeFragment[]) {
```

Comment:
Do you have a suggestion that works without these RelationshipAttributeFragments or similar?

Reply:
I image it in a way that you have a list of attributes (all mustBeAccepted or accepted, see my comment above). Then you check whether there are attributes that have the same values for specified keys. This way, you wouldn't have to create a separate list of RelationshipAttributeFragments.

This is only a high-level thought of me, I haven't checked how this would be implemented in detail. If you see problems with the implementation, feel free to reach out.

Reply:
The problem why I came up with the RelationshipAttributeFragments was that only a RelationshipAttributeQuery and no RelationshipAttribute needs to be provided for the ReadAttributeRequestItem. So I can't just extract RelationshipAttributes (as I could do it with the CreateAttributeRequestItem or ProposeAttributeRequestItem) and then look for duplicates. Also, I need to handle the case that the `owner` of the RelationshipAttribute could be an empty string or the explicit `address` of the recipient. 

I'm thinking about how I could implement the whole logic, including recursions, without using such lists of RelationshipAttribute(Fragment)s and might discuss this with you again.